dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,467 @@
-	// across all of the claims. For example, if you specified "numa", then
-	// this overall claim could only be successfully fulfilled if all of
-	// the included claims could be fulfilled by pools with the same "numa"
-	// attribute value. If we simply set matchAttributes in each claim
-	// separately, then they could be consistent within claims, but
-	// inconsistent across claims. Therefore, we need this additional
-	// constraint.
+// Constraint must have one and only one field set.
+type Constraint struct {
+	// All devices must have this attribute and its value must be the same.
+	//
+	// For example, if you specified "numa.dra.example.com" (a hypothetical example!),
+	// then only devices in the same NUMA node will be chosen.
 	//
 	// +optional
 	// +listType=atomic
-	MatchAttributes []string `json:"matchAttributes,omitempty"`
+	MatchAttribute *string `json:"matchAttribute,omitempty"`
 
-	// Devices contains the actual device claim details, each entry
-	// of which must be sastisfied by selecting one or more devices.
-	// Each entry is either a claim for set of devices, or a list of
-	// prioritized claims for sets of devices, one of which must be
-	// satisfied.
+	// Future extension, not part of the current design:
+	// A CEL expression which compares different devices and returns
+	// true if they match.
 	//
-	// +required
-	// +listType=atomic
-	Devices []DeviceClaimInstance `json:"devices,omitempty"`
+	// Because it would be part of a one-of, old schedulers will not
+	// accidentally ignore this additional, for them unknown match
+	// criteria.
+	//
+	// matcher string
 }
 
-// DeviceClaimInstance captures a group of claims which must all be satisfied,
-// and a group for which one must be sastisfied. If you require multiple
-// groups, one of each may be satsfied, you will need multiple instances.
-type DeviceClaimInstance struct {
-	// Name is used to associate this instance with a specific
-	// container.
-	Name string `json:"name"`
-
-	// Previously, a DeviceClaimInstance resolved to a single
-	// DeviceClaimDetail, so we didn't need matchAttributes here. With
-	// this version, we *could* use matchAttributes here, but most use
-	// cases are likely satisfied by the matchAttributes at the Spec level,
-	// so for now we leave it out of this level.
+// Request is a request for one of many resources required for a claim.
+// This is typically a request for a single resource like a device, but can
+// also ask for one of several different alternatives.
+type Request struct {
+	// Name can be set to enable referencing this request in a pod.spec.containers[].resources.devices
+	// entry, if that is desired.
+	//
+	// Must be a DNS label.
+	//
+	// +optional
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
 
-	// AllOf contains a list of claims, all of which must be satisfied.
-	// be empty.
-	AllOf []DeviceClaimDetail `json:"allOf,omitempty"`
+	*ResourceRequestDetail `json:",inline,omitempty"`
```

Comment:
The separate struct makes sense if we consider adding `OneOf` as an alternative because then we need the same fields here and in that array.

If we inline all fields here, then IMHO we cannot easily move them into a struct later because it changes the protobuf encoding - perhaps this can be handled with manual conversion code between different versions. :shrug: